modules/tinymce/src/core/main/ts/keyboard/Autocompleter.ts

```diff
@@ -0,0 +1,130 @@
+      fireAutocompleterEnd(editor);
+      activeAutocompleter.clear();
+    }
+  };
+
+  const commenceIfNecessary = (context: AutocompleteContext) => {
+    if (!isActive()) {
+      // Create the wrapper
+      AutocompleteTag.create(editor, context.range);
+
+      // store the element/context
+      activeAutocompleter.set({
+        triggerChar: context.triggerChar,
+        matchLength: context.text.length
+      });
+
+      return true;
+    } else {
+      return false;
+    }
+  };
+
+  // This needs to be calculated once things are ready, but the key events must be bound
+  // before `init` or other keydown / keypress listeners will fire first. Therefore,
+  // this is a thunk so that its value is calculated just once when it is used for the
+  // first time, and after that it's value is stored.
+  const getAutocompleters: () => Autocompleters.AutocompleterDatabase = Thunk.cached(() => Autocompleters.register(editor));
+
+  const doLookup = (fetchOptions?: Record<string, any>): Optional<AutocompleteLookupInfo> =>
+    activeAutocompleter.get().map(
+      (ac) => getContext(editor.dom, editor.selection.getRng(), ac.triggerChar)
+        .bind((newContext) => lookupWithContext(editor, getAutocompleters, newContext, fetchOptions))
+    ).getOrThunk(() => lookup(editor, getAutocompleters));
+
+  const load = (fetchOptions?: Record<string, any>) => {
+    doLookup(fetchOptions).fold(
+      cancelIfNecessary,
+      (lookupInfo) => {
+        const wasNecessary = commenceIfNecessary(lookupInfo.context);
+
+        // Wait for the results to return and then display the menu
+        lookupInfo.lookupData.then((lookupData) => {
+          // Lookup the active autocompleter to make sure it's still active, if it isn't then do nothing
+          activeAutocompleter.get().map((ac) => {
+            const context = lookupInfo.context;
+
+            // Ensure the active autocompleter trigger matches, as the old one may have closed
+            // and a new one may have opened. If it doesn't match, then do nothing.
+            if (ac.triggerChar === context.triggerChar) {
+              // close if we haven't found any matches in the last 10 chars
+              if (context.text.length - ac.matchLength >= 10) {
+                cancelIfNecessary();
+              } else {
+                ac.matchLength = context.text.length;
+
+                if (wasNecessary) {
+                  fireAutocompleterStart(editor, { lookupData });
+                } else {
+                  fireAutocompleterUpdate(editor, { lookupData });
+                }
+              }
+            }
+          });
+        });
+      }
+    );
+  };
+
+  editor.editorCommands.addCommand('ReloadAutocompleter', (_ui, value) => {
+    const fetchOptions: Record<string, any> = Type.isObject(value) ? value.fetchOptions : {};
+    load(fetchOptions);
+  });
+
+  editor.editorCommands.addCommand('CloseAutocompleter', () => cancelIfNecessary());
```

Comment:
ETA reduction on these kinds of things are not safe in typescript since if I where to add a boolean to the cancelIfNecessary that has some other purpose than the ui flag then this would break without a compiler warning. Since TS allows you to sick in a function with less arguments in a type hole with more arguments things can go bad. But I changed this anyway.